### PR TITLE
implement cr candidates related

### PR DESCRIPTION
**Pull request by raindust**

### CODE REVIEW

File: cr/state/checkpoint.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+func (k *KeyFrame) DeserializeCodeAddressMap(r io.Reader) (
+	cmap map[string]common.Uint168, err error) {
+	var count uint64
+	if count, err = common.ReadVarUint(r, 0); err != nil {
+		return
+	}
+	cmap = make(map[string]common.Uint168)
+	for i := uint64(0); i < count; i++ {
+		var k string
```
Comment (peterwillcn) [inline cr/state/checkpoint.go]:
```
var k string
```
The variable k is defined outside of the for loop to reduce garbage collection pressure and save memory resources

File: cr/state/checkpoint.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+		if k, err = common.ReadVarString(r); err != nil {
+			return
+		}
+		var v common.Uint168
```
Comment (peterwillcn) [inline cr/state/checkpoint.go]:
```
var v common.Uint168
```
Same as above

File: cr/state/state.go [hunk 3]
```diff
@@ -1,48 +1,150 @@
...
-	// todo complete me
+	KeyFrame
+
+	mtx    sync.RWMutex
+	params *config.Params
 }
 
 func (s *State) GetCandidate(programCode []byte) *Candidate {
-	panic("implement me")
+	did, ok := s.getDIDByCode(programCode)
```
[the remark this answers is not in the document]
Comment (raindust) [reply cr/state/state.go]: fixed, my mistake

File: cr/state/checkpoint_test.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	assert.True(t, keyFrameEqual(frame, frame2))
+}
+
+func keyFrameEqual(first *KeyFrame, second *KeyFrame) bool {
+	for k := range first.Nicknames {
+		if _, ok := second.Nicknames[k]; !ok {
+			return false
+		}
+	}
+
```
Comment (Houshoupei84) [inline cr/state/checkpoint_test.go]: the length of first.CodeDIDMap and second.CodeDIDMap must be equal

File: cr/state/checkpoint_test.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	return candidatesMapEqual(first.PendingCandidates, second.PendingCandidates) &&
+		candidatesMapEqual(first.ActivityCandidates, second.ActivityCandidates) &&
+		candidatesMapEqual(first.CanceledCandidates, second.CanceledCandidates)
+}
+
+func candidatesMapEqual(first map[common.Uint168]*Candidate,
+	second map[common.Uint168]*Candidate) bool {
+	for k, v := range first {
```
Comment (Houshoupei84) [inline cr/state/checkpoint_test.go]: same as above map length

File: cr/state/state_test.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
+package state
+
+import (
+	"github.com/elastos/Elastos.ELA/common"
+	"github.com/stretchr/testify/assert"
+	"testing"
```
Comment (yz89) [inline cr/state/state_test.go]: miss format

Comment (raindust) [reply cr/state/state_test.go]: fixed

File: cr/state/checkpoint_test.go (hunk 4 above)
Comment (raindust) [reply cr/state/checkpoint_test.go]: Nice catch, fixed

File: cr/state/checkpoint_test.go (hunk 5 above)
Comment (raindust) [reply cr/state/checkpoint_test.go]: fixed

**Pull request merged**